### PR TITLE
Synthesize a minimal landing page for combined archives

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://79160385

## Summary

This makes the `docc merge` command synthesize a minimal "landing page" for the combined documentation. 

It also adds two new flags `--synthesized-landing-page-name` and `--synthesized-landing-page-kind` to customize small details about this page. The existence of the pair of these flags is indicted by a new entry in "features.json"

## Dependencies

None

## Testing

- Build documentation for two targets with dependencies. For example: [Inner.doccarchive (zip)](https://github.com/user-attachments/files/16658116/Inner.doccarchive.zip) and 
[Outer.doccarchive (zip)](https://github.com/user-attachments/files/16658118/Outer.doccarchive.zip).

### Default name and kind

- Merge the two archives into a combined archive using
  ```
  swift docc merge /path/to/Inner.doccarchive /path/to/Outer.doccarchive
  ```
- Preview the combined documentation. For example:
  - Run `cd Combined.doccarchive` followed by `ruby -run -ehttpd . -p8080`
  
- Open a browser and navigate to <localhost:8080/documentation/>
  - The landing page should display "Documentation" for the title in the navigator and display name on the page
  - The landing page should display "Package" for the title heading.
  - The landing page should display links to both modules (in alphabetical order)
  Screenshot 2024-08-19 at 11 15 15

  
### Custom name and kind

- Remove the previous combined archive:
  `rm -rf Combined.doccarchive`

- Merge the two archives into a combined archive using, passing a custom name and/or kind
  ```
  swift docc merge \
    /path/to/Inner.doccarchive /path/to/Outer.doccarchive \
    --synthesized-landing-page-name "My Custom Combined Title" \
    --synthesized-landing-page-kind "My Custom Landing Page Kind" 
  ```
- (_unchanged_) Preview the combined documentation. For example:
  - Run `cd Combined.doccarchive` followed by `ruby -run -ehttpd . -p8080`
  
- Open a browser and navigate to <localhost:8080/documentation/>
  - The landing page should display your custom landing page title for the title in the navigator and display name on the page
  - The landing page should display your custom landing page kind for the title heading.
  - (_unchanged_) The landing page should display links to both modules (in alphabetical order)
  Screenshot 2024-08-19 at 11 14 49

### Navigate the combined archive

- While still previewing the combined documentation:
  - Click a few elements in the navigator hierarchy:
    - You should navigate to that page
    - The navigator hierarchy should continue to display the landing page and both modules
  - Click the landing page title at the top of the navigator hierarchy:
    - You should navigate to the landing page 
    - (_unchanged_) The navigator hierarchy should continue to display the landing page and both modules
    
## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- ~[ ] Updated documentation if necessary~ This feature is still experimental and may change

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by sofiaromorales** [comment]
Is there a reason why we are displaying the API collections in a grid format? since this page is synthesised the author will not be able to customise the header image on each tile. Could we consider using a list instead?

Screenshot 2024-08-20 at 10 43 07

**Comment by sofiaromorales** [comment]
Trying to merge two tutorials will raise an error because both contains a table of contents. `Tutorial.doccarchive' and 'Tutorial2.doccarchive' both contain '/data/tutorials/table-of-contents/` 

A possibility to mitigate this is to disambiguate the names if two urls from the merged archives collide.

**Comment by d-ronnqvist** [comment]
> Trying to merge two tutorials will raise an error because both contains a table of contents. `Tutorial.doccarchive' and 'Tutorial2.doccarchive' both contain '/data/tutorials/table-of-contents/`
> 
> A possibility to mitigate this is to disambiguate the names if two urls from the merged archives collide.

That is the expected behavior. It's the developer's responsibility to not name the top-level content they merge the same.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
> Is there a reason why we are displaying the API collections in a grid format? since this page is synthesised the author will not be able to customise the header image on each tile. Could we consider using a list instead?
> 
> 
> 
> Screenshot 2024-08-20 at 10 43 07
> 
> 

Yes, the page looks rather sparse as a list.

The developer authors each modules page image in the markup for that module extension file so they _are_ able to customize that image.

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift [hunk 1]
```diff
@@ -18,7 +18,8 @@ extension Docc {
         public init() {}
         
         public static var configuration = CommandConfiguration(
-            abstract: "Merge a list of documentation archives into a combined archive."
+            abstract: "Merge a list of documentation archives into a combined archive.",
+            usage: "docc merge <archive-path> ... [[--landing-page-catalog <catalog-path>] | [<synthesized-landing-page-options>]] [--output-path <output-path>]"
```
Comment (sofiaromorales) [inline Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift:22]: From https://github.com/swiftlang/swift-docc/pull/1011/files#diff-090d5c64cdfc0f4fdff7b75a5ec658e2eae3edf27ca3795cfba9ef2d9a504822R80 I understand that passing a landing page catalog is not supported yet, right? In that case I think we should remove the landing page catalog parameter from the CLI for now.

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift [hunk 2]
```diff
@@ -94,6 +99,8 @@ extension Docc {
                     guard fileManager.directoryExists(atPath: catalog.path) else {
                         throw ValidationError("No directory exists at '\(catalog.path)'")
                     }
+                    
+                    print("note: Using a custom landing page catalog isn't supported yet. Will synthesize a default landing page instead.")
```
Comment (sofiaromorales) [inline Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift:103]: Why not remove this option for now?

**Review by sofiaromorales — APPROVED** [review]
All minor comments.

A good enhancement that I think it should be part of the PR is giving the developer the possibility to customise the topic section style via a CLI option, instead of using the detailed grid as a fixed value.

**Comment by d-ronnqvist** [comment]
> A good enhancement that I think it should be part of the PR is giving the developer the possibility to customise the topic section style via a CLI option, instead of using the detailed grid as a fixed value.

I added a new flag for this in https://github.com/swiftlang/swift-docc/pull/1011/commits/5be3a6657000b321a87d77bd60158d423a97dd1f

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift (hunk 2 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Merge.swift:104]: See [this comment](https://github.com/swiftlang/swift-docc/pull/1011#discussion_r1723704217)

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Pull request merged**